### ISSUE TITLE
BuildConfig validation allows invalid strategies, causes panics in build controller

### ISSUE THREAD

**Issue by liggitt** [openshift/origin#4447]
Extracted from https://github.com/openshift/origin/pull/4416#issuecomment-136673792

This file creates a BuildConfig object that passes validation, but causes panics when the BuildController tries to instantiate the build.

eg. [kill.yaml](https://gist.githubusercontent.com/nak3/6387633d1e833929cc2d/raw/3ac14d7f4a9569fba7c3ee01daa8ba3d823712e5/killer.yaml)
~~~
oc create -f https://gist.githubusercontent.com/nak3/6387633d1e833929cc2d/raw/3ac14d7f4a9569fba7c3ee01daa8ba3d823712e5/killer.yaml
~~~

- origin.log
~~~
I0901 19:44:25.260717    7279 config_controller.go:38] Running build for BuildConfig default/hello-c
I0901 19:44:25.269988    7279 generator.go:167] Generating Build from BuildConfig: default/hello-c, LastVersion: 0
E0901 19:44:25.271091    7279 util.go:72] Recovered from panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/tmp/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/util.go:66
/tmp/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/util.go:54
/usr/lib/golang/src/runtime/asm_amd64.s:402
/usr/lib/golang/src/runtime/panic.go:387
/usr/lib/golang/src/runtime/panic.go:42
/usr/lib/golang/src/runtime/sigpanic_unix.go:26
/tmp/origin/_output/local/go/src/github.com/openshift/origin/pkg/build/generator/generator.go:237
/tmp/origin/_output/local/go/src/github.com/openshift/origin/pkg/build/generator/generator.go:177
/tmp/origin/_output/local/go/src/github.com/openshift/origin/pkg/build/registry/instantiate/generator/generator.go:35
~~~

Looks like conversion is setting dockerStrategy to an empty struct, so it passes the nil check, but is missing required parameters used during Instantiate.

Validation should prevent that object from getting persisted

**Comment by liggitt** [comment]
The persisted object looks like this:
```
apiVersion: v1
items:
- apiVersion: v1
  kind: BuildConfig
  metadata:
    creationTimestamp: 2015-09-01T13:19:03Z
    labels:
      app: hello-c
    name: hello-c
    namespace: default
    resourceVersion: "367"
    selfLink: /oapi/v1/namespaces/default/buildconfigs/hello-c
    uid: 03d8762a-50ac-11e5-a4ae-3c970e4b7ffe
  spec:
    output:
      to:
        kind: ImageStreamTag
        name: hello-c:latest
    resources: {}
    source:
      git:
        uri: https://github.com/nak3/hello-c.git
      type: Git
    strategy:
      customStrategy:
        from:
          kind: ImageStreamTag
          name: custom-builder-c-make:latest
      dockerStrategy: {}
      type: Docker
    triggers:
    - github:
        secret: 1qBYH35x-XVbYz6NVqWJ
      type: GitHub
    - generic:
        secret: ceoUl6Mu_mWmqv6Qf443
      type: Generic
    - type: ConfigChange
    - imageChange: {}
      type: ImageChange
  status:
    lastVersion: 0
kind: List
metadata: {}
```

**Comment by liggitt** [comment]
We should do the following:
* Sweep for calls to `GetImageStreamForStrategy` and make sure nil is tolerated or handled appropriately
* Update the defaulting function to avoid auto-creating the dockerStrategy (if that's possible while remaining backwards compatible)
* Update validation (if needed) to ensure valid strategy parameters

**Comment by bparees** [comment]
and update the validation logic so that i's invalid to have an ICT with a Docker Strategy that has no From definition.

and update the validation logic to not reject invalid trigger types (related issue)

**Comment by bparees** [comment]
related to https://bugzilla.redhat.com/show_bug.cgi?id=1257576

**Comment by bparees** [comment]
@rhcarvalho here's a summary of alll the things that need to be fixed via this issue:

1) it should be invalid to have a docker buildconfig that has no From and has an ICT with no From.
2) the docker strategy should not be getting automatically added to the buildconfig during conversion (but we need to understand why that was done in the first place before removing it)
3) update the validation logic to not reject "invalid" trigger types (https://bugzilla.redhat.com/show_bug.cgi?id=1259263)

**Comment by rhcarvalho** [comment]
@bparees 

1) changing this can break existing OpenShift deployments. Before #3748 was merged I was able to deploy and use this BC, which now panics (fixing in #4505), and later would become invalid:

```json
{
  "apiVersion": "v1",
  "kind": "BuildConfig",
  "metadata": {
    "name": "chained-builds-app1"
  },
  "spec": {
    "output": {
      "to": {
        "kind": "ImageStreamTag",
        "name": "chained-builds-app1:latest"
      }
    },
    "source": {
      "contextDir": "app",
      "git": {
        "uri": "https://github.com/rhcarvalho/chained-builds-ex.git"
      },
      "type": "Git"
    },
    "strategy": {
      "type": "Docker"
    },
    "triggers": [
      {
        "type": "ImageChange"
      }
    ]
  }
}
```

Are we okay with tightening the validation here?

2) I believe we add an empty `{}DockerBuildStrategy` as a stopgap solution for not checking the possibly nil pointer everywhere. IOW, whenever strategy=docker we make sure the dockerStrategy field is initialized. Similar for other strategies.

3) We/I can do this, but sounds like existing versions already do the aggressive validation, so softening it now wouldn't fix the two BZ mentioned above (https://bugzilla.redhat.com/show_bug.cgi?id=1257576 and https://bugzilla.redhat.com/show_bug.cgi?id=1259263) -- anyway, I think it is a separate matter, unrelated to "BuildConfig validation allows invalid strategies"? (trigger != strategy)

**Comment by bparees** [comment]
1) yes.  the current state today is that the BC causes panics.  I don't see why it would be worse to start rejecting it.

2) hmmm.  i don't love it but i guess i can live with it staying that way.

3) correct, it won't fix the versions in the field, but at least we won't propagate the problem into new versions.  I haven't decided what, if anything, we can do about the versions in the field.  The client tools would have to recognize they were talking to an old version and not define the ConfigChangeTrigger.    The reason I suggested you fix it here is it's also part of validation and you're going to be in the validation code.

**Comment by rhcarvalho** [comment]
1) I don't understand why that template is invalid now and wasn't before. What exactly makes it invalid?

**Comment by bparees** [comment]
1) it was invalid before too.

**Comment by rhcarvalho** [comment]
@bparees updating this issue:

1) I'm not yet buying that it was invalid before. If we build OpenShift as of 12faf376bbb0ec1407d0a06f21bb126d002360bd, which is one commit earlier than 603eab741040fbb06219de02587ca6dce45837b2 (the one that introduced the nil pointer dereference) the template works without errors. Earlier, ImageChangeTriggers did not take a From object reference, and it was valid to have a BC with Docker strategy without a From, in which case the FROM line of the Dockerfile was preserved.
Seems than we introduced `ImageChangeTrigger.From` and this would make `dockerStrategy.From` *sometimes* required, which is a backwards incompatible change.

2) Validation requires that the `dockerStrategy: {}` must be present:
[pkg/build/api/validation/validation.go#L267-L272](https://github.com/openshift/origin/blob/11321c4c08d4b97553b1855c5c9f2f95e64e689b/pkg/build/api/validation/validation.go#L267-L272)
This is the code that initializes the `dockerStrategy` when needed:
[pkg/build/api/v1/conversion.go#L120-L125](https://github.com/openshift/origin/blob/11321c4c08d4b97553b1855c5c9f2f95e64e689b/pkg/build/api/v1/conversion.go#L120-L125)
[pkg/build/api/v1beta3/conversion.go#L127-L132](https://github.com/openshift/origin/blob/11321c4c08d4b97553b1855c5c9f2f95e64e689b/pkg/build/api/v1beta3/conversion.go#L127-L132)

I understand we're not changing this. The fact that the template also defines the `customStrategy` field for a "Docker" strategy build should probably just be ignored.

3) Fixing the validation for a better future: #4556

**Comment by bparees** [comment]
1) it was invalid, it's just that it didn't used to cause an error because we weren't trying to update anything until that commit.  I'm not going to dig into the code history, but it never made sense to have an ICT with no From in conjunction with a Docker strategy with no From.  I don't know exactly what the behavior would have been, but it wasn't "valid" because it makes no sense.  It may not have panicked or thrown an error, but it wasn't valid.  So adding validation that treats it as invalid is the right thing to do.  

tl;dr: If you have a Strategy with no From and an ICT with no From, you've told us "I want my build to be image change triggered, but i'm not going to tell you what image to trigger on", that is invalid and we should be flagging it as such because otherwise we've set the user up for failure, even if it doesn't panic, because their ICT will never fire.


The point being that back when we changed the ICT to not specify a From is really when this problem was introduced (by me).  At that point we(I) should have introduced validation to confirm that if you define an ICT, your Strategy defines a From.  And then when we enabled ICTs to specify their own From, we should have expanded that validation to confirm that either your ICT specifies a From, or your Strategy defines a From.  That the lack of that validation only manifested itself as a panic after additional code changes were introduced (the commit you identified) isn't the key issue.

So at this point we need to fix the validation so no more of these bad BCs get created, and fix the code so that for the existing BCs, we do not panic when we encounter them.

2) yes technically there is no harm in having multiple strategy values populated, as long as the one specified by the Type is properly populated.  Our validation logic should ignore the other strategies (the ones that do not match the Type).

3) :+1:

**Comment by rhcarvalho** [comment]
1) Thanks for the explanation. Can't we possibly populate the `ICT.From` based on the `Dockerfile.FROM`? I'm working on the validation.

2) 
> Our validation logic should ignore the other strategies (the ones that do not match the Type).

It's ok, we already do it as you described.

**Comment by bparees** [comment]
1) we'd have to git clone the repo to get the Dockerfile to do that, that's not something we can do from the places we'd need to do it (the openshift master/controllers).

**Comment by bparees** [comment]
Also the Dockerfile could change (i update my repo with a new Dockerfile, how are you going to know to update the buildconfig?)
